# Concevez une application au service de la santé publique

# 1. Import, paramétrage du notebook et création du DataFrame 'data'

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Modification des paramètres de Pandas pour 
# pouvoir afficher plus de lignes et plus de colonnes.
pd.set_option("max_rows", 200)
pd.set_option("display.max_columns",200)
pd.set_option('display.float_format', lambda x: '%.5f' % x)

In [3]:
data = pd.read_csv('en.openfoodfacts.org.products.csv', encoding='utf8', sep='\t', low_memory=False)

# 2. Infos du DataFrame 'data'

In [4]:
data.tail()

code                                                url  \
1371722    9999999900686  http://world-en.openfoodfacts.org/product/9999...   
1371723       9999999901  http://world-en.openfoodfacts.org/product/9999...   
1371724    9999999910128  http://world-en.openfoodfacts.org/product/9999...   
1371725    9999999990397  http://world-en.openfoodfacts.org/product/9999...   
1371726  999999999999999  http://world-en.openfoodfacts.org/product/9999...   

                            creator   created_t      created_datetime  \
1371722                     kiliweb  1563261470  2019-07-16T07:17:50Z   
1371723                     kiliweb  1585068504  2020-03-24T16:48:24Z   
1371724                     kiliweb  1572513866  2019-10-31T09:24:26Z   
1371725                     kiliweb  1581171613  2020-02-08T14:20:13Z   
1371726  openfoodfacts-contributors  1587222660  2020-04-18T15:11:00Z   

         last_modified_t last_modified_datetime  \
1371722       1588865820   2020-05-07T15:37:00Z   
1371723       1585068505   2020-03-24T16:48:25Z   
1371724       1572513866   2019-10-31T09:24:26Z   
1371725       1582644247   2020-02-25T15:24:07Z   
1371726       1587223187   2020-04-18T15:19:47Z   

                                   product_name generic_name quantity  \
1371722                          Marrons glacés          NaN      NaN   
1371723                                     Scs          NaN      NaN   
1371724  Sandwich club Rillette poisson combava          NaN      NaN   
1371725                                    Fati          NaN    440 g   
1371726                                     NaN          NaN      NaN   

        packaging packaging_tags brands brands_tags      categories  \
1371722       NaN            NaN    NaN         NaN  Marrons glacés   
1371723       NaN            NaN    NaN         NaN             NaN   
1371724       NaN            NaN    NaN         NaN             NaN   
1371725       NaN            NaN    NaN         NaN             NaN   
1371726       NaN            NaN    NaN         NaN             NaN   

                                           categories_tags  \
1371722  en:plant-based-foods-and-beverages,en:plant-ba...   
1371723                                                NaN   
1371724                                                NaN   
1371725                                                NaN   
1371726                                                NaN   

                                             categories_en origins  \
1371722  Plant-based foods and beverages,Plant-based fo...     NaN   
1371723                                                NaN     NaN   
1371724                                                NaN     NaN   
1371725                                                NaN     NaN   
1371726                                                NaN     NaN   

        origins_tags manufacturing_places manufacturing_places_tags labels  \
1371722          NaN                  NaN                       NaN    NaN   
1371723          NaN                  NaN                       NaN    NaN   
1371724          NaN                  NaN                       NaN    NaN   
1371725          NaN                  NaN                       NaN    NaN   
1371726          NaN                  NaN                       NaN    NaN   

        labels_tags labels_en emb_codes emb_codes_tags  \
1371722         NaN       NaN       NaN            NaN   
1371723         NaN       NaN       NaN            NaN   
1371724         NaN       NaN       NaN            NaN   
1371725         NaN       NaN       NaN            NaN   
1371726         NaN       NaN       NaN            NaN   

        first_packaging_code_geo  cities cities_tags purchase_places stores  \
1371722                      NaN     nan         NaN             NaN    NaN   
1371723                      NaN     nan         NaN             NaN    NaN   
1371724                      NaN     nan         NaN             NaN    NaN   
13717

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1371727 entries, 0 to 1371726
Columns: 181 entries, code to carnitine_100g
dtypes: float64(122), int64(2), object(57)
memory usage: 1.8+ GB


Ce DataFrame compte par moins de 181 colonnes.

# 3. Filtrage du DF 'data' avec les produits vendus en France uniquement

L'appel à projet provenant de 'Santé Public France' je filtre les données du DataFrame avec les produits vendu en France uniquement.

## 3.1 Recherche de la variable utilisée sur le site web d'Open Food Fact pour trier les produits par lieux de vente (ici pour la 'France')

Ci-dessous j'effectue différents tests pour identifier la variable utilisée sur le site web pour filtrer les produits par lieux de vente.

En 1er j'effectue une recherche sur le site web en requêtant l'ensemble des produits vendu en France.

Ensuite, j'effectue la même démarche sur le DF 'data' en testant les différentes variables pouvant être utilisées à cette fin.

Enfin, je compare le nombre de produits retournés entre le site web et mes requêtes sur le DF 'data' pour déduire la bonne variable à utiliser pour filtrer les 'produits vendus en France' sur mon Dataframe.

In [6]:
print('Nombre de lignes dans data sans filtre : {}'.format(len(data)))
print('Nombre de lignes dans data avec filtre \'France\' sur \'countries\' : {}'.format(len(data[data['countries'].str.contains('France', na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'countries\' : {}'.format(len(data[data['countries'].str.contains('France', case=False, na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'countries\' OU \'countries_en\' : {}'.format(len(data[(data['countries'].str.contains('France', case=False, na=False) | (data['countries_en'].str.contains('France', case=False, na=False)))])))
print('Nombre de lignes dans data avec filtre \'fr\' sur \'countries\' : {}'.format(len(data[data['countries'].str.contains('fr', case=False, na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'purchase_places\' : {}'.format(len(data[data['purchase_places'].str.contains('France', na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'countries_tags\' : {}'.format(len(data[data['countries_tags'].str.contains('France', na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'countries_en\' : {}'.format(len(data[data['countries_en'].str.contains('france', case=False, na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'countries_tags\' : {}'.format(len(data[data['countries_tags'].str.contains('france', case=False, na=False)])))
print('Nombre de lignes dans data avec filtre \'France\' sur \'ingredients_text\' : {}'.format(len(data[data['ingredients_text'].str.contains('France', na=False)])))

Nombre de lignes dans data sans filtre : 1371727
Nombre de lignes dans data avec filtre 'France' sur 'countries' : 471272
Nombre de lignes dans data avec filtre 'France' sur 'countries' : 530690
Nombre de lignes dans data avec filtre 'France' sur 'countries' OU 'countries_en' : 680758
Nombre de lignes dans data avec filtre 'fr' sur 'countries' : 681278
Nombre de lignes dans data avec filtre 'France' sur 'purchase_places' : 75343
Nombre de lignes dans data avec filtre 'France' sur 'countries_tags' : 0
Nombre de lignes dans data avec filtre 'France' sur 'countries_en' : 680756
Nombre de lignes dans data avec filtre 'France' sur 'countries_tags' : 680756
Nombre de lignes dans data avec filtre 'France' sur 'ingredients_text' : 13586


La variable '**countries_en**' est la variable utilisée sur le site web pour classer les produits par 'pays de vente'.

## 3.2 Filtrage des produits vendu en France uniquement

In [7]:
data = data[data['countries_en'].str.contains('france', case=False, na=False)]
data.head()

code                                                url  \
0         0000000000017  http://world-en.openfoodfacts.org/product/0000...   
1         0000000000031  http://world-en.openfoodfacts.org/product/0000...   
3         0000000000100  http://world-en.openfoodfacts.org/product/0000...   
4  00000000001111111111  http://world-en.openfoodfacts.org/product/0000...   
5         0000000000123  http://world-en.openfoodfacts.org/product/0000...   

                      creator   created_t      created_datetime  \
0                     kiliweb  1529059080  2018-06-15T10:38:00Z   
1                    isagoofy  1539464774  2018-10-13T21:06:14Z   
3                       del51  1444572561  2015-10-11T14:09:21Z   
4  openfoodfacts-contributors  1560020173  2019-06-08T18:56:13Z   
5                     kiliweb  1535737982  2018-08-31T17:53:02Z   

   last_modified_t last_modified_datetime                  product_name  \
0       1561463718   2019-06-25T11:55:18Z              Vitória crackers   
1       1539464817   2018-10-13T21:06:57Z                         Cacao   
3       1444659212   2015-10-12T14:13:32Z   moutarde au moût de raisin    
4       1560020173   2019-06-08T18:56:13Z                       Sfiudwx   
5       1535737986   2018-08-31T17:53:06Z         Sauce Sweety chili 0%   

  generic_name quantity packaging packaging_tags         brands  \
0          NaN      NaN       NaN            NaN            NaN   
1          NaN    130 g       NaN            NaN            NaN   
3          NaN     100g       NaN            NaN  courte paille   
4          NaN    dgesc       NaN            NaN           Watt   
5          NaN      NaN       NaN            NaN            NaN   

     brands_tags                               categories  \
0            NaN                                      NaN   
1            NaN                                      NaN   
3  courte-paille  Epicerie, Condiments, Sauces, Moutardes   
4           watt                                      Xsf   
5            NaN                                      NaN   

                                    categories_tags  \
0                                               NaN   
1                                               NaN   
3  en:groceries,en:condiments,en:sauces,en:mustards   
4                                            fr:xsf   
5                                               NaN   

                          categories_en origins origins_tags  \
0                                   NaN     NaN          NaN   
1                                   NaN     NaN          NaN   
3  Groceries,Condiments,Sauces,Mustards     NaN          NaN   
4                                fr:xsf     NaN          NaN   
5                                   NaN     NaN          NaN   

  manufacturing_places manufacturing_places_tags         labels  \
0                  NaN                       NaN            NaN   
1                  NaN                       NaN            NaN   
3                  NaN                       NaN  Delois france   
4                  NaN                       NaN            NaN   
5                  NaN                       NaN            NaN   

        labels_tags         labels_en emb_codes emb_codes_tags  \
0               NaN               NaN       NaN            NaN   
1               NaN               NaN       NaN            NaN   
3  fr:delois-france  fr:delois-france       NaN            NaN   
4               NaN               NaN       NaN            NaN   
5               NaN               NaN       NaN            NaN   

  first_packaging_code_geo  cities cities_tags purchase_places         stores  \
0                      NaN     nan         NaN             NaN            NaN   
1                      NaN     nan         NaN             NaN            NaN   
3                      NaN     nan         NaN             NaN  courte paille   
4                      NaN     nan         NaN             NaN            NaN   
5    

# 4. Nettoyage du DF 'data'

Cette étape importante vise à optimiser l'exploitation de notre DataFrame.<br>
Les données en entrées étant saisies à la main par les utilisateurs du site web, <br>les informations des produits peuvent être erronées et/ou incomplètes.<br>
De plus, beaucoup d'informations disponibles nous seront inutiles, car inutiles <br>pour notre projet d'application et/ou trop peu renseignés à l'échelle de notre DataFrame.

<u>J'effectue ci-dessous deux principales étapes de nettoyage</u> :
1. Tri et nettoyage des colonnes (ou features) du DF '**data**'

2. Tri et nettoyage des lignes (ou produits)







## 4.1 Visualisation du taux de remplissage des colonnes du DF '**data**'

In [8]:
dataStatsCols = (data.notnull().sum()/len(data)).sort_values(ascending=False)
dataStatsCols

code                                         1.00000
last_modified_datetime                       1.00000
countries_en                                 1.00000
states                                       1.00000
states_tags                                  1.00000
states_en                                    1.00000
countries                                    1.00000
countries_tags                               1.00000
last_modified_t                              1.00000
created_datetime                             1.00000
created_t                                    1.00000
url                                          1.00000
creator                                      1.00000
pnns_groups_2                                0.99947
pnns_groups_1                                0.98592
product_name                                 0.97385
image_small_url                              0.88863
image_url                                    0.88863
energy_100g                                  0

## 4.2 Nettoyage par features (colonnes)

### 4.2.1 Parsage des dates au format yyyymmdd

Je convertis les colonnes '**created_t**' et '**last_modified_t**' en format date puis je reformate la date au format **yyyymmdd** pour pouvoir classer facilement.

In [9]:
data['created_t'] = pd.to_datetime(data['created_t'], unit='s').dt.strftime("%Y%m%d")
data['last_modified_t'] = pd.to_datetime(data['last_modified_t'], unit='s').dt.strftime("%Y%m%d")

#### 4.2.1.1 Suppression des autres colonnes dates (doublon)

In [10]:
data.drop(columns=['created_datetime','last_modified_datetime'], inplace=True)
data.head()

code                                                url  \
0         0000000000017  http://world-en.openfoodfacts.org/product/0000...   
1         0000000000031  http://world-en.openfoodfacts.org/product/0000...   
3         0000000000100  http://world-en.openfoodfacts.org/product/0000...   
4  00000000001111111111  http://world-en.openfoodfacts.org/product/0000...   
5         0000000000123  http://world-en.openfoodfacts.org/product/0000...   

                      creator created_t last_modified_t  \
0                     kiliweb  20180615        20190625   
1                    isagoofy  20181013        20181013   
3                       del51  20151011        20151012   
4  openfoodfacts-contributors  20190608        20190608   
5                     kiliweb  20180831        20180831   

                   product_name generic_name quantity packaging  \
0              Vitória crackers          NaN      NaN       NaN   
1                         Cacao          NaN    130 g       NaN   
3   moutarde au moût de raisin           NaN     100g       NaN   
4                       Sfiudwx          NaN    dgesc       NaN   
5         Sauce Sweety chili 0%          NaN      NaN       NaN   

  packaging_tags         brands    brands_tags  \
0            NaN            NaN            NaN   
1            NaN            NaN            NaN   
3            NaN  courte paille  courte-paille   
4            NaN           Watt           watt   
5            NaN            NaN            NaN   

                                categories  \
0                                      NaN   
1                                      NaN   
3  Epicerie, Condiments, Sauces, Moutardes   
4                                      Xsf   
5                                      NaN   

                                    categories_tags  \
0                                               NaN   
1                                               NaN   
3  en:groceries,en:condiments,en:sauces,en:mustards   
4                                            fr:xsf   
5                                               NaN   

                          categories_en origins origins_tags  \
0                                   NaN     NaN          NaN   
1                                   NaN     NaN          NaN   
3  Groceries,Condiments,Sauces,Mustards     NaN          NaN   
4                                fr:xsf     NaN          NaN   
5                                   NaN     NaN          NaN   

  manufacturing_places manufacturing_places_tags         labels  \
0                  NaN                       NaN            NaN   
1                  NaN                       NaN            NaN   
3                  NaN                       NaN  Delois france   
4                  NaN                       NaN            NaN   
5                  NaN                       NaN            NaN   

        labels_tags         labels_en emb_codes emb_codes_tags  \
0               NaN               NaN       NaN            NaN   
1               NaN               NaN       NaN            NaN   
3  fr:delois-france  fr:delois-france       NaN            NaN   
4               NaN               NaN       NaN            NaN   
5               NaN               NaN       NaN            NaN   

  first_packaging_code_geo  cities cities_tags purchase_places         stores  \
0                      NaN     nan         NaN             NaN            NaN   
1                      NaN     nan         NaN             NaN            NaN   
3                      NaN     nan         NaN             NaN  courte paille   
4                      NaN     nan         NaN             NaN            NaN   
5                      NaN     nan         NaN             NaN            NaN   

   countries countries_tags countries_en  \
0     France      en:france       France   
1     France      en:france       France   
3     France      en:france       France   
4  en:France      en:france       F

### 4.2.2 Suppression des colonnes vides et peu renseignées

Je supprimes les features renseignées à moins de 19% (pas 20% pour conserver les features 'stores' et 'fiber_100g')

In [11]:
data.drop(columns=dataStatsCols[dataStatsCols < 0.19].index, inplace=True)
data.head()

code                                                url  \
0         0000000000017  http://world-en.openfoodfacts.org/product/0000...   
1         0000000000031  http://world-en.openfoodfacts.org/product/0000...   
3         0000000000100  http://world-en.openfoodfacts.org/product/0000...   
4  00000000001111111111  http://world-en.openfoodfacts.org/product/0000...   
5         0000000000123  http://world-en.openfoodfacts.org/product/0000...   

                      creator created_t last_modified_t  \
0                     kiliweb  20180615        20190625   
1                    isagoofy  20181013        20181013   
3                       del51  20151011        20151012   
4  openfoodfacts-contributors  20190608        20190608   
5                     kiliweb  20180831        20180831   

                   product_name quantity packaging packaging_tags  \
0              Vitória crackers      NaN       NaN            NaN   
1                         Cacao    130 g       NaN            NaN   
3   moutarde au moût de raisin      100g       NaN            NaN   
4                       Sfiudwx    dgesc       NaN            NaN   
5         Sauce Sweety chili 0%      NaN       NaN            NaN   

          brands    brands_tags                               categories  \
0            NaN            NaN                                      NaN   
1            NaN            NaN                                      NaN   
3  courte paille  courte-paille  Epicerie, Condiments, Sauces, Moutardes   
4           Watt           watt                                      Xsf   
5            NaN            NaN                                      NaN   

                                    categories_tags  \
0                                               NaN   
1                                               NaN   
3  en:groceries,en:condiments,en:sauces,en:mustards   
4                                            fr:xsf   
5                                               NaN   

                          categories_en         labels       labels_tags  \
0                                   NaN            NaN               NaN   
1                                   NaN            NaN               NaN   
3  Groceries,Condiments,Sauces,Mustards  Delois france  fr:delois-france   
4                                fr:xsf            NaN               NaN   
5                                   NaN            NaN               NaN   

          labels_en         stores  countries countries_tags countries_en  \
0               NaN            NaN     France      en:france       France   
1               NaN            NaN     France      en:france       France   
3  fr:delois-france  courte paille     France      en:france       France   
4               NaN            NaN  en:France      en:france       France   
5               NaN            NaN     France      en:france       France   

                                    ingredients_text  additives_n  \
0                                                NaN          nan   
1                                                NaN          nan   
3  eau graines de téguments de moutarde vinaigre ...      0.00000   
4                                                NaN          nan   
5                                                NaN          nan   

   ingredients_from_palm_oil_n  ingredients_that_may_be_from_palm_oil_n  \
0                          nan                                      nan   
1                          nan                                      nan   
3                      0.00000                                  0.00000   
4                          nan                                      nan   
5                          nan                                      nan   

   nutriscore_score nutriscore_grade  nova_group   pnns_groups_1  \
0               nan              NaN         nan         unknown   
1               nan              NaN         nan         unknown   
3 

### 4.2.3 MAJ du taux de remplissage compte-tenu des suppressions de colonnes

Affichage en **ordre alphabétique** pour **analyse des colonnes inutiles** et **colonnes regroupés par noms**.

In [12]:
dataStatsCols = (data.notnull().sum()/len(data))
dataStatsCols

code                                      1.00000
url                                       1.00000
creator                                   1.00000
created_t                                 1.00000
last_modified_t                           1.00000
product_name                              0.97385
quantity                                  0.34766
packaging                                 0.21946
packaging_tags                            0.21945
brands                                    0.56205
brands_tags                               0.56200
categories                                0.43958
categories_tags                           0.43958
categories_en                             0.43958
labels                                    0.31480
labels_tags                               0.31484
labels_en                                 0.31484
stores                                    0.19317
countries                                 1.00000
countries_tags                            1.00000


### 4.2.4 Analyse et Suppression des colonnes inutiles

Après une analyse du contenu des différentes colonnes, je décide de supprimer les colonnes listées ci-dessous et stockés dans la variable **colsASuppr**.

Liste des colonnes inutiles à supprimer :

In [13]:
colsASuppr = ['url','creator','quantity', 'packaging', 'packaging_tags','brands','brands_tags','categories','categories_tags','categories_en',
              'labels','labels_tags','labels_en','stores','countries','countries_tags','countries_en','ingredients_text','ingredients_from_palm_oil_n',
              'ingredients_that_may_be_from_palm_oil_n','states','states_tags','states_en','main_category','main_category_en','image_url',
              'image_small_url','image_ingredients_url','image_ingredients_small_url','image_nutrition_url','image_nutrition_small_url','additives_n','nova_group']

Suppression des colonnes inutiles :

In [14]:
data.drop(columns=colsASuppr,inplace=True)
data.head()

code created_t last_modified_t  \
0         0000000000017  20180615        20190625   
1         0000000000031  20181013        20181013   
3         0000000000100  20151011        20151012   
4  00000000001111111111  20190608        20190608   
5         0000000000123  20180831        20180831   

                   product_name  nutriscore_score nutriscore_grade  \
0              Vitória crackers               nan              NaN   
1                         Cacao               nan              NaN   
3   moutarde au moût de raisin           18.00000                d   
4                       Sfiudwx               nan              NaN   
5         Sauce Sweety chili 0%               nan              NaN   

    pnns_groups_1         pnns_groups_2  energy-kcal_100g  energy_100g  \
0         unknown               unknown        1569.00000   1569.00000   
1         unknown               unknown               nan          nan   
3  Fat and sauces  Dressings and sauces               nan    936.00000   
4         unknown               unknown               nan          nan   
5         unknown               unknown          88.00000     88.00000   

   fat_100g  saturated-fat_100g  carbohydrates_100g  sugars_100g  fiber_100g  \
0   7.00000             3.08000            70.10000     15.00000         nan   
1       nan                 nan                 nan          nan         nan   
3   8.20000             2.20000            29.00000     22.00000     0.00000   
4       nan                 nan                 nan          nan         nan   
5   0.00000             0.00000             4.80000      0.40000         nan   

   proteins_100g  salt_100g  sodium_100g  nutrition-score-fr_100g  
0        7.80000    1.40000      0.56000                      nan  
1            nan        nan          nan                      nan  
3        5.10000    4.60000      1.81102                 18.00000  
4            nan        nan          nan                      nan  
5        0.20000    2.04000      0.81600                      nan

A ce stade je conserve les colonnes ci-dessous qui me seront utiles dans ce projet.

Le DF 'data' n'est pas encore exploitable.

Il me reste à nettoyer les enregistrements du DataFrame, c'est l'objet de la partie suivante.

## 4.3 Nettoyage par produits (lignes)

Dans cette partie je nettoie le DF 'data' de la façon suivante :
1. Analyse et traitements des données redondantes
2. Analyse et traitement des produits trop peu renseignés
3. Analyse et traitements des données aberrantes
4. Analyse et traitements des données manquantes (hors nutriscore et nutrigrade)

### 4.3.1 Analyse et traitements des données redondantes

#### 4.3.1.1 Recherche des valeurs en doublon en comparants l'ensemble des colonnes disponibles

In [15]:
data.duplicated(subset=None, keep='first')

0          False
1          False
3          False
4          False
5          False
           ...  
1371716    False
1371717    False
1371719    False
1371721    False
1371726    False
Length: 680756, dtype: bool

Il existe 49 lignes exactement en doublon dans le DF '**data**'.

In [16]:
data.duplicated(subset=None, keep='first').value_counts()

False    680704
True         52
dtype: int64

Je supprime les doublons en conservant uniquement les premières occurrences

In [17]:
data = data[~data.duplicated(subset=None, keep='first')]

#### 4.3.1.2 Recherche des valeurs en doublon en comparants toutes les colonnes excepté : **code**,**created_t** et	**last_modified_t**

Les colonnes **code**,**created_t** et	**last_modified_t** peuvent être différentes pour deux produits identiques <br>(par exemple, deux même produits renseignés à des moments différents).

In [18]:
listColsFiltre = data.loc[:,'product_name':].columns
data.duplicated(subset=listColsFiltre, keep='first').value_counts()

False    616058
True      64646
dtype: int64

**59497** doublons trouvés.

Je les supprime du DataFrame '**data**'

In [19]:
data = data[~data.duplicated(subset=listColsFiltre, keep='first')]

#### 4.3.1.3 Recherche des valeurs en doublon en comparants leurs noms

In [20]:
data[data['product_name'].notnull()].duplicated(subset=['product_name'], keep='first').value_counts()

False    458676
True     155965
dtype: int64

Il peut bien entendu exister des produits portant le même nom sans être réellement <br>le même produit ni posséder les mêmes valeurs nutritionnelles.<br>

Je ne supprime donc pas de produits de mon DataFrame sur la base de ce seul indicateur.

#### 4.3.1.4 Recherche des valeurs en doublon en comparants leurs noms et leurs valeurs nutritionnelles [*_100g].

In [21]:
data.duplicated(subset=['product_name','energy-kcal_100g', 'energy_100g', 'fat_100g',
       'saturated-fat_100g', 'carbohydrates_100g', 'sugars_100g', 'fiber_100g',
       'proteins_100g', 'salt_100g', 'sodium_100g'], keep='first').value_counts()

False    608234
True       7824
dtype: int64

Il s'agit ici de réels doublons, je les supprime du DataFrame '**data**'.

In [22]:
data = data[~data.duplicated(subset=['product_name','energy-kcal_100g', 
       'energy_100g', 'fat_100g', 'saturated-fat_100g', 
       'carbohydrates_100g', 'sugars_100g', 'fiber_100g',
       'proteins_100g', 'salt_100g', 'sodium_100g'], keep='first')]

### 4.3.2 Analyse et traitement des produits trop peu renseignés

Un produit avec trop peu d'informations renseignés ne nous est pas utile, <br>je décide de les rechercher et de les supprimer du DataFrame '**data**'.

#### 4.3.2.1 Suppression des produits sans valeurs nutritionnel renseigné (*_100g)

In [23]:
print('Longueur du DF \'data\' avant traitement : {}'.format(len(data)))
listCols100g = data.filter(like='100g', axis=1).columns
# On retire la colonne 'nutrition-score-fr_100g' de la sélection
listCols100g = listCols100g.drop('nutrition-score-fr_100g')
print('Contenue de la variable \'listCols100g\': \n{}'.format(listCols100g))
print('Il y a {} produits qui n\'ont aucune valeur nutritionnel (_100g) \
de renseigné.\nOn les supprime'.format(len(data)-len(data.dropna(subset=listCols100g, how='all'))))
#Il y a 93680 produits qui n'ont aucune valeurs nutritionnel (_100g) de renseigné, je les supprime
data.dropna(subset=listCols100g, how='all',inplace=True)
print('Longueur du DF \'data\' après traitement : {}'.format(len(data)))

Longueur du DF 'data' avant traitement : 608234
Contenue de la variable 'listCols100g': 
Index(['energy-kcal_100g', 'energy_100g', 'fat_100g', 'saturated-fat_100g',
       'carbohydrates_100g', 'sugars_100g', 'fiber_100g', 'proteins_100g',
       'salt_100g', 'sodium_100g'],
      dtype='object')
Il y a 88106 produits qui n'ont aucune valeur nutritionnel (_100g) de renseigné.
On les supprime
Longueur du DF 'data' après traitement : 520128


#### 4.3.2.2 Suppression des produits sans nom de produit renseigné ('product_name' vide)

In [24]:
data[data['product_name'].isnull()]
print('Je supprime les {} produits sans nom renseigné'.format(len(data[data['product_name'].isnull()])))
data = data[data['product_name'].notnull()]

Je supprime les 1365 produits sans nom renseigné


### 4.3.3 Analyse et traitements des données aberrantes

Je décide de supprimer les produits contenants au moins 1 valeur aberrante.<br>
<u>J'identifie comme aberrante les valeurs qui correspondent à au moins l'un des critères suivants</u> :
1. Les produits avec des valeurs >100 dans au moins l'une des colonnes [*_100g]
2. Les produits avec des valeurs négatives dans au moins l'une des colonnes [*_100g]
3. Les produits avec carbohydrates_100g > sugars_100g
4. Les produits avec saturated-fat_100g > fat_100g
5. Les produits avec la valeur 'energy_100g' anormalement élevé


In [25]:
#Re-calcul du taux de remplissage
dataStatsCols = (data.notnull().sum()/len(data)).sort_values(ascending=False)
dataStatsCols

code                      1.00000
pnns_groups_2             1.00000
created_t                 1.00000
last_modified_t           1.00000
product_name              1.00000
pnns_groups_1             0.99670
energy_100g               0.98837
proteins_100g             0.98332
saturated-fat_100g        0.98089
sugars_100g               0.98071
fat_100g                  0.97914
carbohydrates_100g        0.97873
salt_100g                 0.96245
sodium_100g               0.96245
energy-kcal_100g          0.90961
nutrition-score-fr_100g   0.42674
nutriscore_grade          0.42673
nutriscore_score          0.42673
fiber_100g                0.24508
dtype: float64

#### 4.3.3.1 Suppression des produits avec des valeurs >100 dans au moins l'une des colonnes [*_100g]

On retire préalablement les colonnes qui peuvent légitimement contenir des valeurs supérieures à 100.

Les colonnes restantes sont enregistrées dans la variable '**newListCols100g**'.

In [26]:
colsAExclure = ['energy-kcal_100g','energy_100g','nutrition-score-fr_100g']
newListCols100g = [ele for ele in listCols100g if ele not in colsAExclure] 
print(newListCols100g) # Affichage des colonnes qui ne doivent pas contenir de valeur supérieure à 100.

['fat_100g', 'saturated-fat_100g', 'carbohydrates_100g', 'sugars_100g', 'fiber_100g', 'proteins_100g', 'salt_100g', 'sodium_100g']


Je crée une fonction qui filtre le DF donnée en entrée en retirant tous <br>
les enregistrements qui contiennent au moins une valeur supérieure <br>
à 100 parmi les colonnes à contrôler.

In [27]:
def filtreValeursAberantes(df, cols, value, inf=True):
  '''Prend un df en entrée et retourne un df filtré
Si inf vaut True alors seules les lignes qui ont toutes 
leurs valeurs inférieures ou égales à 'value', dans 
l'ensemble des colonnes 'cols', sont conservées
Sinon, seules les lignes qui ont toutes 
leurs valeurs supérieur ou égales à 'value', dans 
l'ensemble des colonnes 'cols', sont conservées'''
  #Initialisation de la liste d'index à retourner
  if inf :
    listeIndexValide = df[cols].fillna(0).apply(lambda row: row.name if (row<=value).all() else None, axis=1)
  else:
    listeIndexValide = df[cols].fillna(0).apply(lambda row: row.name if (row>=value).all() else None, axis=1)
  listeIndexValide.dropna(inplace=True)
  listeIndexValide = [int(i) for i in listeIndexValide]
  return df[df.index.isin(listeIndexValide)]

In [28]:
print('Nombre de produits avant filtrage des valeurs aberrantes : {}'.format(len(data)))
data = filtreValeursAberantes(data,newListCols100g,100, True)
print('Nombre de produits après filtrage des valeurs aberrantes : {}'.format(len(data)))

Nombre de produits avant filtrage des valeurs aberrantes : 518763
Nombre de produits après filtrage des valeurs aberrantes : 518696


#### 4.3.3.2 Suppression des produits avec des valeurs négatives dans au moins l'une des colonnes [*_100g]

Il ne doit pas exister de valeurs négatives dans les colonnes indiquant <br>la quantité en grammes d’un nutriment pour 100 grammes du produit.

Les produits renseignant des valeurs négatives sur au moins l'un <br>des nutriments est supprimé.

In [29]:
print('Nombre de produits avant filtrage des valeurs aberrantes : {}'.format(len(data)))
data = filtreValeursAberantes(data,newListCols100g,0, False)
print('Nombre de produits après filtrage des valeurs aberrantes : {}'.format(len(data)))

Nombre de produits avant filtrage des valeurs aberrantes : 518696
Nombre de produits après filtrage des valeurs aberrantes : 518695


#### 4.3.3.3 Suppression des produits avec carbohydrates_100g > sugars_100g

Les carbohydrates appartiennent au groupe des sucres. <br>Leur quantité pour une produit donnée ne peut donc pas être supérieur au '**sucre**'.

Les produits dont la quantité de carbohydrates_100g <br>est supérieur à sugars_100g sont supprimés.

In [30]:
print('Suppression de {} produits'.format(len(data[data.sugars_100g > data.carbohydrates_100g])))
data = data[data.sugars_100g <= data.carbohydrates_100g]

Suppression de 179 produits


#### 4.3.3.4 Suppression des produits avec saturated-fat_100g > fat_100g

Les graisses saturées sont une portion des graisses. <br>Elles ne peuvent donc pas leur être supérieur en quantité.

Les produits dont la quantité de saturated-fat_100g<br> 
est supérieur à fat_100g sont supprimés.

In [31]:
print('Suppression de {} produits'.format(len(data[data['saturated-fat_100g'] > data.fat_100g])))
data = data[data['saturated-fat_100g'] <= data.fat_100g]

Suppression de 189 produits


#### 4.3.3.5 Traitement des colonnes energy_100g et energy-kcal_100g

J'émet l'hypothèse que les deux colonnes font doublons.<br>
J'étudie leur contenu, et tente de les nettoyer de leurs valeurs aberrantes.

In [32]:
data[['energy_100g']].sort_values(by='energy_100g',ascending=False).head(10)

energy_100g
784276  345671211.00000
1151912     34044.00000
491608      19305.00000
1136971     14242.00000
1045293      9544.00000
1123031      9364.00000
1124549      9347.00000
1126723      9192.00000
1126722      9192.00000
325214       9192.00000

Au moins 1 valeur de la colonne 'energy_100g' est totalement aberrante et sera supprimée.

In [33]:
data[['code','product_name','energy-kcal_100g','energy_100g']].sort_values(by='energy_100g',ascending=False).head(10)

code                            product_name  \
784276   3661112082529                         Farce à légumes   
1151912  8003250001096                           Olio Di Oliva   
491608        25168242  barres noisettes et chocolat equitable   
1136971     7889514788                   Pure black sesame oil   
1045293  5690576270188         Traditional Icelandic Chocolate   
1123031  7622210079657                                 Suchard   
1124549  7622210521811                              Dairy Milk   
1126723  7622300331283                                  Moreni   
1126722  7622300331269         Moreni Chocolate Wafer / Peanut   
325214   0728229129417                          Sweets & Beets   

         energy-kcal_100g     energy_100g  
784276                nan 345671211.00000  
1151912        3448.00000     34044.00000  
491608        19305.00000     19305.00000  
1136971       14242.00000     14242.00000  
1045293        9544.00000      9544.00000  
1123031        9364.00000      9364.00000  
1124549        9347.00000      9347.00000  
1126723        9192.00000      9192.00000  
1126722        9192.00000      9192.00000  
325214         9192.00000      9192.00000

Les valeurs semblent identiques dans les deux colonnes.<br>
Je vérifie cela ci-dessous à l'échelle du DataFrame :

In [34]:
print(len(data[data.energy_100g == data['energy-kcal_100g']]))
print(len(data[data.energy_100g < data['energy-kcal_100g']]))
print(len(data[data.energy_100g > data['energy-kcal_100g']]))
print(len(data[data.energy_100g == (data['energy-kcal_100g'])*0.23])) # 1 kiloJoule = 0.23 kiloCalorie
print(len(data[data['energy-kcal_100g'] > 1000]))

449089
7016
9142
10503
256839


1. 61% des valeurs des deux colonnes sont identiques.
2. 0.14% des valeurs de la colonne energy_100g ont un rapport de 0.23 <br>avec la colonne energy-kcal_100g, ce qui indique que ces valeurs, <br>de la colonne energy_100g, sont renseigné en kilojoules.
3. Le reste des valeurs renseignés, environ 39%, <br>n'ont pas de rapport identifié entre les deux.

In [35]:
data[data['energy-kcal_100g'] > 1000]

code created_t last_modified_t  \
0        0000000000017  20180615        20190625   
18        000000000114  20200126        20200208   
19       0000000001151  20180925        20180925   
21       0000000001281  20180205        20180523   
46       0000000004206  20190218        20190218   
...                ...       ...             ...   
1371708  9999990060556  20190809        20190809   
1371710  9999990060631  20180925        20180925   
1371715  9999991149090  20180220        20181220   
1371717       99999988  20190103        20190716   
1371719  9999999004360  20190121        20190520   

                                   product_name  nutriscore_score  \
0                              Vitória crackers               nan   
18                                Chocolate n 3               nan   
19                &quot;Baguette Lyonnais&quot;               nan   
21                           Tarte noix de coco          14.00000   
46       Saucisse au chou fumé BBC SV (Morteau)               nan   
...                                         ...               ...   
1371708                 Cailloux coeur chocolat               nan   
1371710                    Cailloux aux amandes               nan   
1371715                             Riz parfumé           0.00000   
1371717                                  Boules               nan   
1371719                          Minis beignets          15.00000   

        nutriscore_grade         pnns_groups_1       pnns_groups_2  \
0                    NaN               unknown             unknown   
18                   NaN               unknown             unknown   
19                   NaN               unknown             unknown   
21                     d         Sugary snacks  Biscuits and cakes   
46                   NaN               unknown             unknown   
...                  ...                   ...                 ...   
1371708              NaN               unknown             unknown   
1371710              NaN               unknown             unknown   
1371715                b  Cereals and potatoes             Cereals   
1371717              NaN               unknown             unknown   
1371719                d               unknown             unknown   

         energy-kcal_100g  energy_100g  fat_100g  saturated-fat_100g  \
0              1569.00000   1569.00000   7.00000             3.08000   
18             2439.00000   2439.00000  44.00000            28.00000   
19             1213.00000   1213.00000   9.40000             4.50000   
21             1594.00000   1594.00000  22.00000            15.50000   
46             4033.00000   4033.00000  25.00000             9.70000   
...                   ...          ...       ...                 ...   
1371708        1690.00000   1690.00000   0.00000             0.00000   
1371710        1979.00000   1979.00000  19.00000            15.00000   
1371715        1469.00000   1469.00000   0.50000             0.00000   
1371717        1556.00000   1556.00000  11.00000             1.30000   
1371719        1393.00000   1393.00000  20.20000             6.00000   

         carbohydrates_100g  sugars_100g  fiber_100g  proteins_100g  \
0                  70.10000     15.00000         nan        7.80000   
18                 30.00000     27.00000         nan        2.10000   
19                 41.00000      2.00000         nan       12.50000   
21                 27.30000     21.90000     4.40000        4.60000   
46                  1.00000      0.10000         nan       20.00000   
...                     ...          ...         ...            ...   
1371708            97.00000     93.00000         nan        0.00000   
1371710            64.00000     63.00000         nan        9.00000   
1371715            80.00000      0.00000         nan        7.00000   
1371717            58.00000     54.00000         nan        6.30000   
1371719            30.90000     10.40000         nan        7.40000   

     

La feature '**energy-kcal_100g**' me semble anormalement mal renseigné. <br>Il est incohérent d'avoir autant de produits qui ont plus de 1000 kcal pour 100g.

Le nutella a par exemple une valeur de 'seulement' 539kcal/100g.

La colonne '**energy_100g**' ne doit pas renseigner des kcal (kilocalorie) mais de kj (kilojoule).

Je suspecte la colonne '**energy-kcal_100g**' d'être majoritairement renseigné <br>avec des valeurs en joule. D'ailleurs beaucoup de valeurs sont identiques <br>entre les 2 colonnes '**energy-kcal_100g**' et '**energy_100g**' alors que 1kcal = 4,184 kjoule.

A ce stade, je considère la colonne '**energy_100g**' mieux renseignés que '**energy-kcal_100g**'.

Je renseigne donc la colonne '**energy-kcal_100g**' à partir des données <br>de la colonne '**energy_100g**' selon le calcul suivant : energy-kcal_100g = energy_100g*0.239006

In [36]:
data['energy-kcal_100g'] = (data['energy_100g']*0.239006)

In [37]:
data[['energy-kcal_100g','energy_100g']].head(10)

energy-kcal_100g  energy_100g
0          375.00041   1569.00000
3          223.70962    936.00000
5           21.03253     88.00000
6           59.99051    251.00000
13          32.02680    134.00000
14         129.06324    540.00000
17         222.03657    929.00000
18         582.93563   2439.00000
19         289.91428   1213.00000
20         218.92950    916.00000

Je supprime maintenant les valeurs aberrantes en supprimant <br>les produits qui ont leur '**energy_100g**' anormalement élevé.

Après analyse, un apport en energie de plus de 8000kJ pour 100g me semble vraiment excessif.<br>
Je décide de supprimer les produits ayant une valeur supérieur à 8000 dans la colonne '**energy_100g**'

La nature aléatoire de cette problématique m'empêche <br>malheureusement d'automatiser la recherche/suppression des valeurs aberrantes non legitimes.

In [38]:
data.sort_values(by='energy_100g',ascending=False).head(30)

code created_t last_modified_t  \
784276                     3661112082529  20171224        20190624   
1151912                    8003250001096  20170627        20200517   
491608                          25168242  20181010        20200227   
1136971                       7889514788  20200323        20200323   
1045293                    5690576270188  20181115        20191223   
1123031                    7622210079657  20190221        20200508   
1124549                    7622210521811  20180802        20200418   
1126723                    7622300331283  20180920        20190701   
1126722                    7622300331269  20180418        20190701   
325214                     0728229129417  20180222        20200111   
496986                     2600692028474  20180914        20180914   
1072754                    6490144260820  20190325        20190325   
238059   0173481712110809151907301018324  20190116        20190116   
1055327                    5908230515538  20181104        20191216   
1140355                    8000300369939  20190202        20190202   
1062482                    6111031004449  20190224        20200406   
61569                      0034000442652  20190211        20200424   
976764                     5052004062466  20190207        20200502   
955182                     5000128605113  20190609        20200504   
478489                          24036412  20170310        20191123   
402742                          10602294  20191211        20191211   
439281                          20891077  20180620        20200428   
1110197                    7613037027198  20200516        20200516   
979083                     5054402919816  20181108        20200308   
1052081                    5901414204495  20180922        20200124   
597712                     3254560075955  20170215        20191030   
959138                     5000221602606  20190522        20191211   
245272                     0202459022309  20190622        20190622   
1159385                    8007900000255  20170319        20191027   
457656                     2240328044746  20181130        20200425   

                                      product_name  nutriscore_score  \
784276                             Farce à légumes          30.00000   
1151912                              Olio Di Oliva           6.00000   
491608      barres noisettes et chocolat equitable               nan   
1136971                      Pure black sesame oil          11.00000   
1045293            Traditional Icelandic Chocolate          36.00000   
1123031                                    Suchard               nan   
1124549                                 Dairy Milk          29.00000   
1126723                                     Moreni               nan   
1126722            Moreni Chocolate Wafer / Peanut               nan   
325214                              Sweets & Beets               nan   
496986                            mélange mexicain               nan   
1072754                 Sablés pépites de chocolat               nan   
238059                                   CROC CAFE               nan   
1055327                          Kabanos exklusive          30.00000   
1140355             Gran Blocco Fondente Extra 50%               nan   
1062482                                       Oreo               nan   
61569                 Reese's Peanut Butter Bar XL               nan   
976764                                  Digestives               nan   
955182   Sea Salt & Chardonnay Wine Vinegar Crisps               nan   
478489              Miel Toutes Fleurs Tamara 500g               nan   
402742     Foie gras de canard entier du Sud Ouest          24.00000   
439281                                  Digestives               nan   
1110197                         Knacki 100% poulet          20.00000   
979083              Tesco Fruit Shortcake Biscuits               nan   
1052081                           KRAKUSKI Kakaowe               

Un apport en energie de plus de 8000kJ pour 100g me semble vraiment excessif.<br>
Je décide de supprimer les produits ayant une valeur supérieur à 8000 dans la colonne '**energy_100g**'

In [40]:
data = data[data['energy_100g'] < 8000]

In [ ]:
data.sort_values(by='energy_100g',ascending=False).head(10)

### 4.3.4 Analyse et traitements des données manquantes (hors nutriscore_score et nutriscore_grade)

La complétion des colonnes **nutriscore_score** et **nutriscore_grade** <br>
seront traités dans une section dédiée où je ferai appel au **Machine Learning**.

Pour la complétion des autres colonnes, je choisi d'utiliser <br>**la moyenne des valeurs disponible** comme valeur de substitution.

Pour que ce choix d'utiliser la moyenne reste cohérent, j'appliquerai <br>la moyenne à chaque produit en fonction de la **catégorie** à laquelle il appartient.

Pour le choix des catégories je m'appuie sur les modalités de la colonne **pnns_group_1**.

<u>Voici le processus que j'applique ci-dessous</u> :
1. Identification des catégories de pnns_groups_1
2. Nettoyage par suppression les produits qui <br>ont comme valeur dans pnns_groups_1 : **NaN** et '**unknown**'. 
3. Calcul de la moyenne de chaque features par catégories.
4. Stockage de ces valeurs dans un tableau
5. Création d'une fonction pour compléter les valeurs manquantes <br>de chaque produit en fonction de la feature et de la catégorie du produit à compléter.

#### 4.3.4.1 Analyse et traitement de la colonne 'pnns_groups_1' (représente les catégories de produits)

In [ ]:
data.pnns_groups_1.value_counts()

<u>DataFrame représentant le taux de remplissage du DF '**data**' si la colonne '**pnns_groups_1**'</u> :


1.   n'a pas de restriction
2.   est différent de 'unknown'
3.   est différent de 'unknow' **ET** n'est pas Null






In [ ]:
d = {'pnns_groups_1=all': (data.notnull().sum()/len(data)), 
     'pnns_groups_1!=unknow': (data[data['pnns_groups_1'] != 'unknown'].notnull().sum()\
                               /len(data[data['pnns_groups_1'] != 'unknown'])),
     'pnns_groups_1=(!=unknow&notnull)': (data[(data.pnns_groups_1 != 'unknown') \
                                            & (data.pnns_groups_1.notnull())].notnull().sum()\
                                       /len(data[(data.pnns_groups_1 != 'unknown') \
                                                 & (data.pnns_groups_1.notnull())]))}

pd.DataFrame(data=d)

#### 4.3.4.2 Filtrage de 'data' avec les données de la colonne 'pnns_groups_1' Non Null ET différent de 'unknown'

In [ ]:
data = data[(data.pnns_groups_1.notnull()) & (data.pnns_groups_1 != 'unknown')]

#### 4.3.4.3 Liste des catégories de produits

*On va utiliser cette liste pour renseigner les données manquantes*

In [ ]:
data.pnns_groups_1.value_counts()

Nombre de produits restant dans le DF '**data**' :

In [ ]:
len(data[(data.pnns_groups_1 != 'unknown') & (data.pnns_groups_1.notnull())])

#### 4.3.4.4 Définition des fonctions de remplissage des données manquantes

In [ ]:
def calcMeanByCateg(df,cols,categ):
  '''Retourne un DF contenant les moyennes 
  calculées de chaque 'cols' classé selon les modalités de la colonne 'categ' 
  * df est le DataFrame 
  * cols corresponds à une liste de noms de colonnes
  * categ correspond au nom d'une colonne'''
  dictValRef = {}
  for cat in df[categ].unique():
    dictValRef.update({cat: df[cols][df[categ] == cat].mean()})
  return pd.DataFrame(dictValRef).T





In [ ]:
def fillMissingValues(df, cols, categ):
  '''Renseigne les valeurs manquantes des colonnes **cols** 
  du DataFrame 'df' à partir des moyennes de ces colonnes.
  Les valeurs sont renseignées par catégories selon les modalités de la colonne **categ**'''

  for categorie in df[categ].unique():
    df.loc[df[categ]==categorie, cols] = df[df[categ]==categorie][cols].fillna(df[df[categ] == categorie][cols].mean())
  return df

#### 4.3.4.5 Affichage du tableau de correspondance (catégorie de la colonne 'pnns_groups_1"  / <br>Moyenne des colonnes contenant *._100g )

In [ ]:
calcMeanByCateg(data, listCols100g, 'pnns_groups_1')

#### 4.3.4.6 Remplissage des données manquantes des colonnes [*_100g] (à l'exception de 'nutrition-score-fr_100g') <br>à partir de leurs moyennes et en fonction des modalités de la colonne 'pnns_groups_1'

In [ ]:
data = fillMissingValues(data, listCols100g, 'pnns_groups_1')
data.head()

### 4.3.5 Taux de remplissage des colonnes du DF 'data' (MAJ)


In [ ]:
dataStatsCols = (data.notnull().sum()/len(data)).sort_values(ascending=False)
dataStatsCols

Toutes les colonnes à l'exception de celles représentants le **nutrition_score** sont maintenant intégralement renseignés.<br>
Dans la partie suivante, je ferai appel à différents algorithme de **Machine Learning** pour compléter les dernières colonnes <br>faisant références au **nutrition_score**.

# 5. Machine Learning

Dans cette partie, j'utilise le Machine Learning pour m'aider à compléter intelligemment <br>les features '**nutrition-score-fr_100g**', '**nutriscore_grade**' et '**nutriscore_score**'.

Les notes (lettre ou chiffre) attribuées par le nutriscore sont définis en fonction <br>
des valeurs nutritionnelles des produits.<br>
L'idée est donc d'entraîner les algorithmes de ML à déterminer intelligemment <br>
les nutriscore manquants, en apprenant à partir des nutriscore déjà renseignés, <br>
et des valeurs nutritionnels correspondantes.

Pour l'ensemble des algorithmes utilisés ici, je me baserai sur la bibliothèque '**scikit-learn**'.<br>
Les algorithmes utilisés sont de types '**supervisés**'.

<u>J'utiliserai 3 algorithmes différents</u> :
1. 2 algorithmes basés sur les regressions lineaires (**LinearRegression** et **KNeighborsRegressor**) <br>
pour déterminer les valeurs continues qui constitues la feature '**nutrition-score-fr_100g**'.<br>
Les 2 algorithmes testés seront comparés et je sélectionnerai le plus performant des deux.
2. 1 algorithme basé sur la classification (**KNeighborsClassifier**), afin de déterminer des valeurs <br>
discrètes comme celles renseignés dans '**nutriscore_grade**'

Pour les algorithmes basés sur la méthode des k plus proches voisins (KNN), j'utilise également <br>
la fonction **GridSearchCV()** qui me permet d'optimiser le choix des hyperparamètres.<br>
De plus, je termine mon étude des algorithmes KNN par l'affichage de la **Confusion Matrix** et de la **Learning Curve**.


## 5.1 Taux de remplissage du DF 'data'

In [ ]:
data.head()

In [ ]:
dataStatsCols = (data.notnull().sum()/len(data)).sort_values(ascending=False)
dataStatsCols

La colonne '**nutrition-score-fr_100g**' est un peu mieux renseigné que la colonne '**nutriscore_score**'. 

Je décide donc d'utiliser '**nutrition-score-fr_100g**' dans la suite de cette partie et <br>d'effacer la colonne '**nutriscore_score**' du DF '**data**'.

In [ ]:
data = data.drop(columns=['nutriscore_score']) # Suppression de la colonne 'nutriscore_score'

## 5.2 Régression Linéaire

Dans cette partie, j'utilise un algorithme de Régression Linéaire pour <br>déterminer les valeurs manquantes de la colonne '**nutrition-score-fr_100g**'.<br>
Comme il s'agit d'un algorithme supervisé, je devrai séparer le dataframe <br>avec d'une part les features et les labels (ici '**nutrition-score-fr_100g**').

Je commence par créer un nouveau dataframe '**dataRL**' <br>débarrassé des valeurs null de la colonne '**nutrition-score-fr_100g**'.

En effet l'algorithme va devoir apprendre en devinant la valeur nutriscore <br>des produits en fonction de leur **features**.
Durant la **phrase d'apprentissage**, <br>l'algorithme compare les valeurs devinées avec les valeurs réels (label, <br>ici '**nutrition-score-fr_100g**').<br>
Pour que l'algorithme puisse fonctionner correctement, il faut  que <br>toutes les valeurs de nutriscore (labels) soient connues. <br>Il est donc nécessaire de supprimer les valeurs null de la colonne '**nutrition-score-fr_100g**'.

In [ ]:
# Création d'un DataFrame sans valeur null pour la colonne ''nutrition-score-fr_100g'
print('Taux de remplissage de \'nutrition-score-fr_100g\' dans \'data\':', 
      data['nutrition-score-fr_100g'].notnull().sum()/len(data))
dataRL = data[data['nutrition-score-fr_100g'].notnull()]
print('Taux de remplissage de \'nutrition-score-fr_100g\' dans \'data\':', 
      dataRL['nutrition-score-fr_100g'].notnull().sum()/len(dataRL))

Je sélectionne maintenant les colonnes features d'un côté <br>et la colonne label correspondant de l'autre.

Par souci de lecture et par convention, <br>je nomme les deux DF par une seule lettre :
1. Le DF contenant les features se nomme '**X**'
2. Le DF contenant les labels se nomme '**y**'

In [ ]:
# Sélection des features
X = dataRL[listCols100g]
y = dataRL['nutrition-score-fr_100g']

In [ ]:
print(X.shape)
print(y.shape)

La taille des DF est de 204 664 lignes.

Les features sont de dimension 10 (10 colonnes) <br>et les labels doivent être eux de dimension 1.

Les DF '**X**' et '**y**' sont maintenant divisé en 2 parties chacune.
1. Le **train set** sert à l'**apprentissage** de l'algorithme
2. Le **test set** sert à **évaluer les performances de l'algorithme** une fois entrainé

Il est courant d'utiliser **80%** des données pour le **train set** et **20%** pour le **test set**.

In [ ]:
# Création du train set et du test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

J'importe maintenant l'algorithme à utiliser puis <br>je lui communique les données d'entrainement via la méthode **fit()**.

Une fois l'entrainement terminé, j'évalue ses résultats <br>avec les données de test via la méthode **score()**.

In [ ]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)
model.score(X_test, y_test)

L'algorithme présente un taux de réussite de **61.7%** sur les données de test.

## 5.3 KNN Regressor

Le même processus est réalisé ici avec l'algorithme **KNN Regressor**.<br>
A la différence de l'algorithme **LinearRegression**, l'algorithme **KNN Regressor** <br>accepte en entrée plusieurs hyperparamètre qui vont influer sur le taux de réussite en sortie.<br>
Pour m'aider à faire le meilleur choix des hyperparamètres, j'utiliserai la méthode **GridSearchCV**.

Nous réutilisons le même DF '**dataRL**' utilisé précédemment ainsi <br>que la même méthodologie de création des données d'entrainement et de test.

In [ ]:
print(dataRL['nutrition-score-fr_100g'].notnull().sum()/len(dataRL))
dataRL.head()

In [ ]:
# Création des features
X = dataRL[listCols100g]
# Création des labels
y = dataRL['nutrition-score-fr_100g']

In [ ]:
# Création du train set et du test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

print(X_train.shape)
print(X_test.shape)

Je réalise un premier essai avec l'ensemble des valeurs des hyperparamètres par défaut :

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
model = KNeighborsRegressor()
model.fit(X_train, y_train)
model.score(X_test, y_test)

J'obtiens un score d'un peu plus de **86%** de réussite sur les données de test.<br>
Le choix de l'algorithme **KNN Regressor**, même non optimisé, me donne un <br>résultat amélioré d'un peu plus de **24%** par rapport à l'algorithme **LineaRegression**.

Je cherche maintenant à optimiser mon algorithme en testant automatiquement, <br>grâce à la méthode GridSeachCV, la valeur des différents **hyperparamètres** <br>et en effectuant une **Cross-Validation**.

La **Cross-Validation** est un technique qui permet à notre algorithme <br>d'apprendre et de tester ses résultats sur différentes partie du jeux de donnée.<br>
Si lors de la phase d'apprentissage l'algorithme utilise **80%** du jeu de donnée <br>et se test sur les **20%** restant, la **Cross-Validation** permettra de répéter <br>le processus en alternant le choix des données utilisé dans le **train-test** <br>et le **test-set** jusqu'à ce que toutes les données ait été utilisé à la fois <br>dans le **train-test et le test-set**, éliminant ainsi les erreurs liées au choix, <br>aléatoire, des données à inclure dans le **train-set** et le **test-set**.

In [ ]:
# Test des différentes valeurs des hyperparamètres et Cross-Validation avec GridSearchCV
from sklearn.model_selection import GridSearchCV
# Dictionnaire contenant les différents hyperparamètres à régler 
# ainsi que chaque valeur à tester pour ces hyperparamètres.
param_grid = {'n_neighbors': np.arange(1,20),
              'metric': ['euclidean', 'manhattan'],
              'weights': ['uniform', 'distance']}
grid = GridSearchCV(KNeighborsRegressor(), param_grid, cv=5)

grid.fit(X_train, y_train)

On affiche maintenant les meilleurs valeurs des hyperparamètres testés :

In [ ]:
print('Meilleurs paramètres:',grid.best_params_)
print('Meilleur score:', grid.best_score_)

Sauvegarde du modèle avec les meilleurs paramètres :

In [ ]:
model = grid.best_estimator_

Test de notre modèle sur les données de test :

In [ ]:
model.score(X_test, y_test)

Notre modèle est maintenant optimisé et nous obtenons un taux de réussite <br>de plus de **88,6%** soit **2,6%** de mieux que notre modèle non optimisé <br>ou **26,6%** de mieux que l'algorithme **LinearRegression**.

### Learning Curve

Je cherche à savoir si la quantité de donnée contenu dans mon DataFrame <br>est suffisante et si la qualité de mon modèle pourrait être amélioré par <br>l'ajout de nouvelles données.

Pour répondre à cette question j'utilise la **Learning Curve** qui me permettra <br>de visualiser la **performance de mon modèle** en fonction de la quantité <br>de données contenu dans mon **validation-set**.

L'apport de nouvelles données peut-être une opération couteuse en temps <br>et/ou financièrement. Il est important de savoir s'il est nécessaire d'ajouter <br>de nouvelles données ou si au contraire cela constitue une perte de temps et/ou d'argent.

In [ ]:
from sklearn.model_selection import learning_curve
N, train_score, val_score = learning_curve(model, X_train, y_train, 
                                           train_sizes = np.linspace(0.1, 1.0, 10), cv=5)
plt.plot(N, train_score.mean(axis=1), label='train')
plt.plot(N, val_score.mean(axis=1), label='validation')
plt.xlabel('train_sizes')
plt.legend()

Comme nous pouvons l'observer, la courbe d'apprentissage <br>n'a pas encore atteint son point d'aplatissement et les résultats <br>de l'algorithme pourrait être amélioré par l'apport de nouvelles données.

### Remplissage des valeurs manquantes calculées grâce au KNN Regressor

Le modèle étant maintenant optimisé et entrainé, je l'utilise pour deviner <br>les valeurs manquantes de ma colonne '**nutrition-score-fr_100g**' grâce à la méthode **predict()**.

In [ ]:
# Remplissage des valeurs manquantes calculées grâce au KNN Regressor
data['nutrition-score-fr_100g'] = data.apply(lambda row: 
            model.predict(np.array(row[listCols100g]).reshape(1,10)) \
            if pd.isnull(row['nutrition-score-fr_100g']) \
            else row['nutrition-score-fr_100g'] ,axis=1)

Vérification du taux de remplissage de la colonne '**nutrition-score-fr_100g**'

In [ ]:
len(data['nutrition-score-fr_100g'])/len(data)

## 5.4 KNN Classifier

J'utilise maintenant l'algorithme **KNeighborsClassifier** pour compléter <br>la colonne '**nutriscore_grade**'.<br>
La colonne '**nutriscore_grade**' contient des valeurs discrètes, pour cette <br>raison j'utilise un algorithme de classification.

<u>Le processus est pratiquement le même que pour l'utilisation de l'algorithme **KNeighborsRegressor**</u> :
1. *Nouveau* : Notre algorithme fonctionne avec des valeurs numériques uniquement. <br>Je commence par remplacer les lettres '**a**' à '**e**' contenu dans la colonne <br>'**nutriscore_grade**' par les valeurs numériques correspondantes allant de '**1**' à '**5**'
2. Sélection des features (X) et des labels (y)
3. Séparation du jeu de donnée en train-set / test-set
4. Test de l'algorithme avec les valeurs d'hyperparamètres par défaut
5. Cross-Validation et recherche des meilleurs réglages des hyperparamètres avec **GridSearchCV**
6. Enregistrement des meilleurs paramètres dans notre modèle
7. Evaluation finale du modèle
8. *Nouveau* : Affichage de la **Confusion Matrix** (affiche les erreurs de classification du modèle)
9. Affichage de la learning Curve
10. Renseignement de la colonne '**nutriscore_grade**' à l'aide de notre modèle entrainé

### 5.4.1 Sélection des features et des labels et <br>remplacement des valeur de 'nutriscore_grade' <br>par des valeurs numériques

Création d'un DataFrame sans valeur null pour la colonne ''nutriscore_grade'

In [ ]:
print(data['nutriscore_grade'].notnull().sum()/len(data))
dataML = data[data['nutriscore_grade'].notnull()]
print(dataML['nutriscore_grade'].notnull().sum()/len(dataML))

J'effectue le remplacement des valeurs de '**nutriscore_grade**' directement dans le DF contenant les labels.

In [ ]:
# Sélection des features
X = dataML[listCols100g]
# Sélection des labels
y = dataML['nutriscore_grade'].replace(['a','b','c','d','e'],[1,2,3,4,5])

### 5.4.2 Séparation du jeu de donnée en train-set / test-set

In [ ]:
# Création du train set et du test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

print(X_train.shape)
print(X_test.shape)

### 5.4.3 Test de l'algorithme avec les valeurs d'hyperparamètre par défaut

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier()
model.fit(X_train, y_train)
model.score(X_test, y_test)

Le modèle non optimisé nous donne un résultat d'environ 73,8% de réussite.

### 5.4.4 Cross-Validation et recherche des meilleurs réglages des hyperparamètres avec GridSearchCV

In [ ]:
# Test des différentes valeurs des hyperparamètres et Cross-Validation avec GridSearchCV
from sklearn.model_selection import GridSearchCV
# Dictionnaire contenant les différents hyperparamètres à régler 
# ainsi que chaque valeur à tester pour ces hyperparamètres.
param_grid = {'n_neighbors': np.arange(1,20),
              'metric': ['euclidean', 'manhattan'],
              'weights': ['uniform', 'distance']}
grid = GridSearchCV(KNeighborsClassifier(), param_grid, cv=5)

grid.fit(X_train, y_train)

In [ ]:
print('Meilleurs paramètres:',grid.best_params_)
print('Meilleur score:', grid.best_score_)

### 5.4.5 Enregistrement et tests des meilleurs paramètres dans notre modèle

In [ ]:
# Sauvegarde du modèle avec les meilleurs paramètres
model = grid.best_estimator_

In [ ]:
# Test de notre modèle sur les données de test
model.score(X_test, y_test)

Le modèle optimisé à un taux de réussite de plus de 78.1%

### 5.4.6 Confusion Matrix

La ***Matrice de confusion*** nous permet de visualiser les erreurs de prédiction du modèle

In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, model.predict(X_test))

On constate par exemple que le modèle a classé correctement <br>
dans la catégorie '**a**' du champs 'nutriscore_grade' **4377** valeurs, 
<br>mais a classé à tort : 
- **718** valeurs dans la catégorie '**b**'
- **348** valeurs dans la catégorie '**c**'
- **88** valeurs dans la catégorie '**d**'
- **32** valeurs dans la catégorie '**e**'

### 5.4.7 Learning Curve

In [ ]:
from sklearn.model_selection import learning_curve
N, train_score, val_score = learning_curve(model, X_train, y_train, 
                                           train_sizes = np.linspace(0.1, 1.0, 10), cv=5)
plt.plot(N, train_score.mean(axis=1), label='train')
plt.plot(N, val_score.mean(axis=1), label='validation')
plt.xlabel('train_sizes')
plt.legend()

On constate que la courbe d'apprentissage n'a pas atteint son point d'écrasement 
<br>et peut encore être amélioré avec l'apport de données supplémentaires.

### 5.4.8 Renseignement des valeurs manquantes pour nutriscore_grade

Je crée ici une fonction qui appel la méthode '**predict**' de notre modèle <br>puis retourne la lettre correspondant au chiffre retourné par la méthode '**predict**' via l'utilisation d'un dictionnaire.

In [ ]:
def predict_nutriscore_grade(model, features):
  label_predict = model.predict(features)
  mapping = {1:'a', 2:'b', 3:'c', 4:'d', 5:'e'}
  return mapping[label_predict.item()]

In [ ]:
# Remplissage de la colonne 'nutriscore_grade'
data['nutriscore_grade'] = data.apply(lambda row: 
            predict_nutriscore_grade(model, np.array(row[listCols100g]).reshape(1,10)) \
            if pd.isnull(row['nutriscore_grade']) \
            else row['nutriscore_grade'] ,axis=1)

In [ ]:
# Vérification du taux de remplissage de la colonne 'nutriscore_grade'
len(data['nutriscore_grade'])/len(data)

# 6. Renommage des colonnes et sauvegarde du DF '**data**'

In [ ]:
data.columns

In [ ]:
data.rename(columns={'nutriscore_grade': 'nutrigrade', 'nutrition-score-fr_100g': 'nutriscore'}, inplace=True)

In [ ]:
data.columns

Je sauvegarde le DF '**data**'' dans un fichier *Pickle* pour l'utiliser ensuite dans la suite de notre projet.

In [ ]:
import pickle
pickle.dump(data, open('data_clean', 'wb'))

# Fin de la partie nettoyage du DataFrame '**data**